include/AST/Decl/FunctionDecl.hpp

```diff
@@ -10,62 +11,106 @@
     FunctionDecl(
-        SourceLocation location, ASTNode *parent, std::string name,
-        glu::types::FunctionTy *type, llvm::SmallVector<Param> params
+        SourceLocation location, ASTNode *parent, llvm::StringRef name,
+        glu::types::FunctionTy *type, llvm::ArrayRef<ParamDecl> const params
     )
         : DeclBase(NodeKind::FunctionDeclKind, location, parent)
         , _name(std::move(name))
         , _type(type)
-        , _params(std::move(params))
         , _body(CompoundStmt(location, {}))
+        , _numParams(params.size())
     {
+        std::uninitialized_copy(
+            params.begin(), params.end(),
+            this->template getTrailingObjects<ParamDecl>()
+        );
         _body.setParent(this);
     }
 
+public:
+    friend TrailingParams;
+
+    /// @brief Static method to create a new FunctionDecl.
+    /// @param alloc The allocator used to create the FunctionDecl.
+    /// @param location The source location of the function declaration.
+    /// @param parent The parent AST node.
+    /// @param name The name of the function.
+    /// @param type The type of the function.
+    /// @param params A vector of Param objects representing the parameters of
+    /// @return Returns a new FunctionDecl.
+    static FunctionDecl *create(
+        llvm::BumpPtrAllocator &alloc, SourceLocation location, ASTNode *parent,
+        llvm::StringRef name, glu::types::FunctionTy *type,
+        llvm::ArrayRef<ParamDecl> const params
+    )
+    {
+        auto totalSize = totalSizeToAlloc<ParamDecl>(params.size());
+        void *mem = alloc.Allocate(totalSize, alignof(FunctionDecl));
+        return new (mem)
+            FunctionDecl(location, parent, std::move(name), type, params);
+    }
+
     /// @brief Getter for the name of the function.
     /// @return Returns the name of the function.
-    std::string getName() const { return _name; }
+    llvm::StringRef getName() const { return _name; }
 
     /// @brief Getter for the type of the function.
     /// @return Returns the type of the function.
     glu::types::FunctionTy *getType() const { return _type; }
 
     /// @brief Getter for the parameters of the function.
-    /// @return Returns a vector of Param objects representing the parameters of
-    /// the function.
-    llvm::ArrayRef<Param> getParams() const { return _params; }
+    /// @return Returns the parameters.
+    llvm::ArrayRef<ParamDecl> getParams() const
+    {
+        return { this->template getTrailingObjects<ParamDecl>(), _numParams };
+    }
+
+    /// @brief Getter of the index of a parameter.
+    /// @param name The name of the parameter to retrieve.
+    /// @return Returns the asked parameter index.
+    std::optional<size_t> getParamIndex(llvm::StringRef name) const
+    {
+        auto params = this->getParams();
+
+        auto searchedParam = std::find_if(
+            params.begin(), params.end(),
+            [&name](ParamDecl const &param) { return param.getName() == name; }
```

Comment:
name is a StringRef, copy is cheap
```suggestion
            [name](ParamDecl const &param) { return param.getName() == name; }
```